### DISCUSSION TITLE
Next steps for COLMAP

### DISCUSSION THREAD

**Discussion by sarlinpe** [colmap/colmap#2380] [General]
Hello everyone,

I want to give an update on what future changes we are planning for COLMAP. We want to make it:

1. **More research-friendly**, i.e. more modular, easier to extend and to hack

2. **More robust**, by leveraging recent research, for example, on learning-based matching, self-calibrating minimal solvers, etc.

3. **Faster**, with more parallelization when possible

4. **Cover a broader set of use cases**, including multi-sensor reconstruction (multi-camera rigs, spherical cameras, GPS, depth), long videos, etc.

We think that progressively moving the higher-level logic to Python will make these easier to achieve, while keeping a performant core C++ library. To get there, we've derived a first set of smaller milestones:

- [x] Tighter integration of pycolmap into colmap https://github.com/colmap/colmap/pull/2367
- [x] Use exception to propagate errors (more Python-friendly) https://github.com/colmap/colmap/pull/2376
- [x] Disentangle the controllers from the threading (@B1ueber2y is taking this) and expose them to Python
- [x] Rely on PoseLib for most solvers https://github.com/colmap/colmap/pull/2288
  - [ ] Would benefit from benchmarking solvers that estimate the camera calibration
- [ ] Two-view verification: support the least-squares refinement of E/H/F matrices via Ceres
- [ ] Create and benchmark a database format based on HDF5 as alternative (and later replacement) to SQL
- [ ] Design and implement the serialization of reconstructions via protocol buffer, as alternative (and later replacement) to the custom bin/text format
- [ ] Switch some internal data types (images, keypoints) to plain Eigen matrices for better interoperability with Python
- [x] More extensive coverage of the Python bindings: database, bitmap, estimators
- [x] Automated documentation generation for the Python bindings
- [x] Proper & tested integration of multi-camera rigs into the pipeline (to later support spherical images via tiling)
- [ ] Proper & tested integration of GPS into the pipeline (maybe a rework of https://github.com/colmap/colmap/pull/1409)

It would be great to get feedback from the community (especially power users) on whether these goals align with their use cases. If some folks have the bandwidth to help on some of these points, you are very much welcome to contribute. We will then be able to tackle deeper changes.

Thanks!

cc @ahojnnes @tsattler @vlarsson @mihaidusmanu @Phil26AT @arjunkarpur @B1ueber2y @Zador-Pataki

**Comment by grafpez** [comment]
I'd love to see multi-camera rigs supported via GUI. I'm not a coder and there is no good other alternatives. This is hampering what I would consider good potential progress to better and faster 3DGS pipelines

**Comment by bashtavenko** [comment]
@sarlinpe 

> Design and implement the serialization of reconstructions via protocol buffer, as alternative (and later replacement) to the custom bin/text format

Are you going to build this in-house or it can be done by contributors? If later, Is there an issue / feature description?
